### ISSUE TITLE
CI for Valgrind

### ISSUE THREAD

**Issue by albinahlback** [flintlib/flint2#1290]
What kind of output is one looking for? What CFLAGS are required? I have never run Valgrind more than once, and I am not familiar with its output. I suppose it is as simple as to run `make valgrind` and check if it fails. @fredrik-johansson @thofma

**Comment by thofma** [comment]
I think I used valgrind only twice in my life and it was full of false positives, so not helpful at all. But I guess there a configuration options to make it useful.

**Comment by fredrik-johansson** [comment]
I use valgrind extensively. Usually I run the individual program I'm interested in, which is as simple as ``valgrind <command>``. No recompilation or special ``CFLAGS`` required.

Quick tutorial how to track down memory leaks.

If everything is fine, I get something like this:

```
fredrik@mordor:~/src/flint2$ valgrind build/fmpz/test/t-add
==2694160== Memcheck, a memory error detector
==2694160== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==2694160== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==2694160== Command: build/fmpz/test/t-add
==2694160== 
add....PASS
==2694160== 
==2694160== HEAP SUMMARY:
==2694160==     in use at exit: 0 bytes in 0 blocks
==2694160==   total heap usage: 154,089 allocs, 154,089 frees, 3,245,670 bytes allocated
==2694160== 
==2694160== All heap blocks were freed -- no leaks are possible
==2694160== 
==2694160== For lists of detected and suppressed errors, rerun with: -s
==2694160== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

If I insert a memory leak in ``fmpz_add``, I get output like this:

```
fredrik@mordor:~/src/flint2$ valgrind build/fmpz/test/t-add
==2694742== Memcheck, a memory error detector
==2694742== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==2694742== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==2694742== Command: build/fmpz/test/t-add
==2694742== 
add....PASS
==2694742== 
==2694742== HEAP SUMMARY:
==2694742==     in use at exit: 511,536 bytes in 9,398 blocks
==2694742==   total heap usage: 172,889 allocs, 163,491 frees, 3,851,294 bytes allocated
==2694742== 
==2694742== LEAK SUMMARY:
==2694742==    definitely lost: 208,896 bytes in 3 blocks
==2694742==    indirectly lost: 302,640 bytes in 9,395 blocks
==2694742==      possibly lost: 0 bytes in 0 blocks
==2694742==    still reachable: 0 bytes in 0 blocks
==2694742==         suppressed: 0 bytes in 0 blocks
==2694742== Rerun with --leak-check=full to see details of leaked memory
==2694742== 
==2694742== For lists of detected and suppressed errors, rerun with: -s
==2694742== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

Any nonzero value of "still reachable", "possibly lost", "indirectly lost" or "definitely lost" indicate a memory leak. If I run ``valgrind`` with more flags, I get more info:

```
fredrik@mordor:~/src/flint2$ valgrind --leak-check=full --show-reachable=yes build/fmpz/test/t-add
==2694747== Memcheck, a memory error detector
==2694747== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==2694747== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==2694747== Command: build/fmpz/test/t-add
==2694747== 
add....PASS
==2694747== 
==2694747== HEAP SUMMARY:
==2694747==     in use at exit: 511,536 bytes in 9,398 blocks
==2694747==   total heap usage: 172,889 allocs, 163,491 frees, 3,851,294 bytes allocated
==2694747== 
==2694747== 16 bytes in 1 blocks are indirectly lost in loss record 1 of 18
==2694747==    at 0x4848899: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x586F99D: __gmp_default_allocate (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x587F896: __gmpz_init2 (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x4A3D055: _fmpz_new_mpz (fmpz.c:105)
==2694747==    by 0x4A3D4A8: _fmpz_promote (fmpz.c:211)
==2694747==    by 0x4A3D4A8: _fmpz_promote (fmpz.c:207)
==2694747==    by 0x4A4C997: fmpz_randtest_unsigned (randtest.c:56)
==2694747==    by 0x4A4CA65: fmpz_randtest (randtest.c:23)
==2694747==    by 0x1093C0: main (t-add.c:42)
==2694747== 
==2694747== 40 bytes in 1 blocks are indirectly lost in loss record 2 of 18
==2694747==    at 0x484DCD3: realloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x586F9E8: __gmp_default_reallocate (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886151: __gmpz_realloc (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5876838: __gmpz_add_ui (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x4A31B9C: fmpz_add (add.c:43)
==2694747==    by 0x1093EE: main (t-add.c:47)
==2694747== 
==2694747== 80 bytes in 2 blocks are indirectly lost in loss record 3 of 18
==2694747==    at 0x484DCD3: realloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x586F9E8: __gmp_default_reallocate (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886151: __gmpz_realloc (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x58765BC: __gmpz_add (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x4A31C1E: fmpz_add (add.c:52)
==2694747==    by 0x1093EE: main (t-add.c:47)
==2694747== 
==2694747== 1,088 bytes in 34 blocks are indirectly lost in loss record 4 of 18
==2694747==    at 0x484DCD3: realloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x586F9E8: __gmp_default_reallocate (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886151: __gmpz_realloc (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886E97: __gmpz_rrandomb (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x4A4C9AF: fmpz_randtest_unsigned (randtest.c:59)
==2694747==    by 0x4A4CA65: fmpz_randtest (randtest.c:23)
==2694747==    by 0x109699: main (t-add.c:126)
==2694747== 
==2694747== 1,152 bytes in 36 blocks are indirectly lost in loss record 5 of 18
==2694747==    at 0x484DCD3: realloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x586F9E8: __gmp_default_reallocate (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886151: __gmpz_realloc (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886E97: __gmpz_rrandomb (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x4A4C9AF: fmpz_randtest_unsigned (randtest.c:59)
==2694747==    by 0x4A4CA65: fmpz_randtest (randtest.c:23)
==2694747==    by 0x1097CF: main (t-add.c:171)
==2694747== 
==2694747== 1,944 bytes in 61 blocks are indirectly lost in loss record 6 of 18
==2694747==    at 0x484DCD3: realloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x586F9E8: __gmp_default_reallocate (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886151: __gmpz_realloc (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886E97: __gmpz_rrandomb (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x4A4C9AF: fmpz_randtest_unsigned (randtest.c:59)
==2694747==    by 0x4A4CA65: fmpz_randtest (randtest.c:23)
==2694747==    by 0x1097BE: main (t-add.c:170)
==2694747== 
==2694747== 2,264 bytes in 71 blocks are indirectly lost in loss record 7 of 18
==2694747==    at 0x484DCD3: realloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x586F9E8: __gmp_default_reallocate (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886151: __gmpz_realloc (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886E97: __gmpz_rrandomb (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x4A4C9AF: fmpz_randtest_unsigned (randtest.c:59)
==2694747==    by 0x4A4CA65: fmpz_randtest (randtest.c:23)
==2694747==    by 0x1096AA: main (t-add.c:127)
==2694747== 
==2694747== 3,888 bytes in 109 blocks are indirectly lost in loss record 8 of 18
==2694747==    at 0x484DCD3: realloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x586F9E8: __gmp_default_reallocate (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886151: __gmpz_realloc (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5888738: __gmpz_sub_ui (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x4A31BD4: fmpz_add (add.c:42)
==2694747==    by 0x1096CE: main (t-add.c:132)
==2694747== 
==2694747== 3,912 bytes in 111 blocks are indirectly lost in loss record 9 of 18
==2694747==    at 0x484DCD3: realloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x586F9E8: __gmp_default_reallocate (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886151: __gmpz_realloc (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5888738: __gmpz_sub_ui (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x4A31BE7: fmpz_add (add.c:31)
==2694747==    by 0x1097F3: main (t-add.c:176)
==2694747== 
==2694747== 4,264 bytes in 120 blocks are indirectly lost in loss record 10 of 18
==2694747==    at 0x484DCD3: realloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x586F9E8: __gmp_default_reallocate (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886151: __gmpz_realloc (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5888738: __gmpz_sub_ui (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x4A31BD4: fmpz_add (add.c:42)
==2694747==    by 0x1097F3: main (t-add.c:176)
==2694747== 
==2694747== 4,456 bytes in 126 blocks are indirectly lost in loss record 11 of 18
==2694747==    at 0x484DCD3: realloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x586F9E8: __gmp_default_reallocate (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886151: __gmpz_realloc (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5888738: __gmpz_sub_ui (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x4A31BE7: fmpz_add (add.c:31)
==2694747==    by 0x1096CE: main (t-add.c:132)
==2694747== 
==2694747== 9,336 bytes in 265 blocks are indirectly lost in loss record 12 of 18
==2694747==    at 0x484DCD3: realloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x586F9E8: __gmp_default_reallocate (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886151: __gmpz_realloc (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5876838: __gmpz_add_ui (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x4A31B9C: fmpz_add (add.c:43)
==2694747==    by 0x1096CE: main (t-add.c:132)
==2694747== 
==2694747== 9,776 bytes in 275 blocks are indirectly lost in loss record 13 of 18
==2694747==    at 0x484DCD3: realloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x586F9E8: __gmp_default_reallocate (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886151: __gmpz_realloc (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5876838: __gmpz_add_ui (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x4A31B9C: fmpz_add (add.c:43)
==2694747==    by 0x1097F3: main (t-add.c:176)
==2694747== 
==2694747== 110,528 (69,632 direct, 40,896 indirect) bytes in 1 blocks are definitely lost in loss record 14 of 18
==2694747==    at 0x4848899: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x49A6DED: flint_malloc (memory_manager.c:111)
==2694747==    by 0x4A3CF01: _fmpz_new_mpz (fmpz.c:78)
==2694747==    by 0x4A3D4A8: _fmpz_promote (fmpz.c:211)
==2694747==    by 0x4A3D4A8: _fmpz_promote (fmpz.c:207)
==2694747==    by 0x4A31C0B: fmpz_add (add.c:49)
==2694747==    by 0x1097F3: main (t-add.c:176)
==2694747== 
==2694747== 129,304 bytes in 4,056 blocks are indirectly lost in loss record 15 of 18
==2694747==    at 0x484DCD3: realloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x586F9E8: __gmp_default_reallocate (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886151: __gmpz_realloc (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x58765BC: __gmpz_add (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x4A31C1E: fmpz_add (add.c:52)
==2694747==    by 0x1097F3: main (t-add.c:176)
==2694747== 
==2694747== 131,120 bytes in 4,127 blocks are indirectly lost in loss record 16 of 18
==2694747==    at 0x484DCD3: realloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x586F9E8: __gmp_default_reallocate (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x5886151: __gmpz_realloc (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x58765BC: __gmpz_add (in /usr/local/lib/libgmp.so.10.4.1)
==2694747==    by 0x4A31C1E: fmpz_add (add.c:52)
==2694747==    by 0x1096CE: main (t-add.c:132)
==2694747== 
==2694747== 200,080 (69,632 direct, 130,448 indirect) bytes in 1 blocks are definitely lost in loss record 17 of 18
==2694747==    at 0x4848899: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x49A6DED: flint_malloc (memory_manager.c:111)
==2694747==    by 0x4A3CF01: _fmpz_new_mpz (fmpz.c:78)
==2694747==    by 0x4A3D4A8: _fmpz_promote (fmpz.c:211)
==2694747==    by 0x4A3D4A8: _fmpz_promote (fmpz.c:207)
==2694747==    by 0x4A4C997: fmpz_randtest_unsigned (randtest.c:56)
==2694747==    by 0x4A4CA65: fmpz_randtest (randtest.c:23)
==2694747==    by 0x1093C0: main (t-add.c:42)
==2694747== 
==2694747== 200,928 (69,632 direct, 131,296 indirect) bytes in 1 blocks are definitely lost in loss record 18 of 18
==2694747==    at 0x4848899: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==2694747==    by 0x49A6DED: flint_malloc (memory_manager.c:111)
==2694747==    by 0x4A3CF01: _fmpz_new_mpz (fmpz.c:78)
==2694747==    by 0x4A3D4A8: _fmpz_promote (fmpz.c:211)
==2694747==    by 0x4A3D4A8: _fmpz_promote (fmpz.c:207)
==2694747==    by 0x4A31C0B: fmpz_add (add.c:49)
==2694747==    by 0x1096CE: main (t-add.c:132)
==2694747== 
==2694747== LEAK SUMMARY:
==2694747==    definitely lost: 208,896 bytes in 3 blocks
==2694747==    indirectly lost: 302,640 bytes in 9,395 blocks
==2694747==      possibly lost: 0 bytes in 0 blocks
==2694747==    still reachable: 0 bytes in 0 blocks
==2694747==         suppressed: 0 bytes in 0 blocks
==2694747== 
==2694747== For lists of detected and suppressed errors, rerun with: -s
==2694747== ERROR SUMMARY: 3 errors from 3 contexts (suppressed: 0 from 0)
```

This points to locations in the code where memory was allocated without subsequently being freed. Note that this is not always the location of the bug. This is particularly an issue with ``fmpz`` because we recycle our allocated ``mpz`` values (``fmpz``-related memory leaks can be easier to debug in ``single`` mode because of this). For example, in the above output, besides the actual point of leak in ``fmpz_add``, but we also get reports from ``fmpz_randtest`` etc. which is where some ``mpz`` just happened to be allocated in the first place.

Valgrind will also detect use of uninitialized data, invalid reads or writes, etc. Those tend to be more straightforward to debug because you usually get the exact location in the code where the error occurs.

You can also use ``valgrind --tool=callgrind`` and ``valgrind --tool=cachegrind`` as a profiling tool.

**Comment by albinahlback** [comment]
How long does it take compared to simply running the all tests? Perhaps we can only do it for merges into `trunk`?

**Comment by fredrik-johansson** [comment]
I don't know about the whole test suite but for example with ``FLINT_TEST_MULTIPLIER=0.1``, ``make valgrind MOD=fmpz`` takes 200x longer than ``make check MOD=fmpz``.

**Comment by albinahlback** [comment]
oh...

**Comment by albinahlback** [comment]
Alright, then it's not feasible.

**Comment by thofma** [comment]
OK, thanks. It will be a bit tricky in Nemo. We can call `flint_cleanup_master` when julia exists, but it is done *before* the individual objects are finalized/cleared. Thanks for the info.
